[extractor/rtbf] Update extractor

elyse0 opened the pull request:
<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### Description of your *pull request* and other information

</details>

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

For its latest videos, RTBF has started using RedBee as its service provider but old videos still use the previous api, so we fallback to RedBee only when no formats were found.

It's important to note that previously we could download videos without account, but for the latest we'll need to use cookies 🍪 or username and password.

The new RedBee extractor comes mostly from the ParliamentLiveUK extractor, since it already had most of the logic needed, it was just missing support for different auth methods, like Gigya for RTBF.

### Verbose log
```bash
[debug] Command-line config: ['https://www.rtbf.be/auvio/detail_formule-1-grand-prix-de-france?id=2920922', '--cookies-from-browser', 'firefox', '-F', '--verbose']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05ef6] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: 30dbd3d4f
[debug] Python 3.8.10 (CPython 64bit) - Linux-5.4.0-122-generic-x86_64-with-glibc2.29 (glibc 2.31)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.0.1-8-g11eff7739a-20220712 (setts), ffprobe n5.0.1-8-g11eff7739a-20220712, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "/home/amish/.mozilla/firefox/ckczlffq.default-release/cookies.sqlite"
[Cookies] Extracted 2703 cookies from firefox
[debug] Proxy map: {}
[debug] [RTBF] Extracting URL: https://www.rtbf.be/auvio/detail_formule-1-grand-prix-de-france?id=2920922
[RTBF] 2920922: Downloading webpage
[RTBF] 2920922: Downloading webpage
[RTBF] 2920922: Downloading JSON metadata
[RTBF] 2920922: Downloading JSON metadata
[RTBF] 2920922: Downloading JSON metadata
[RTBF] 2920922: Downloading MPD manifest
[RTBF] 2920922: Downloading m3u8 information
[RTBF] 2920922: Downloading ISM manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 2920922:
ID                      EXT  RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC        VBR ACODEC      ABR MORE INFO
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
audio-aacl-128-Français m3u8 audio only     │                  m3u8  │ audio only        unknown         [fr] Français
130                     m4a  audio only     │ ~114.26MiB  130k m3u8  │ audio only        mp4a.40.2  128k
448                     mp4  480x270     25 │ ~393.75MiB  448k m3u8  │ avc1.4D4015  448k video only
766                     mp4  640x360     25 │ ~673.24MiB  766k m3u8  │ avc1.4D401E  766k video only
1826                    mp4  960x540     25 │ ~  1.57GiB 1826k m3u8  │ avc1.4D401F 1826k video only
```

### Other video
```bash
[debug] Command-line config: ['--verbose', 'https://www.rtbf.be/auvio/detail_la-belgique-criminelle?id=2920492', '--cookies-from-browser', 'firefox']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05ef6] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: ad047aead
[debug] Python 3.8.10 (CPython 64bit) - Linux-5.4.0-122-generic-x86_64-with-glibc2.29 (glibc 2.31)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.0.1-8-g11eff7739a-20220712 (setts), ffprobe n5.0.1-8-g11eff7739a-20220712, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, secretstorage-2.3.1, sqlite3-2.6.0, websockets-10.3
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "/home/amish/.mozilla/firefox/ckczlffq.default-release/cookies.sqlite"
[Cookies] Extracted 2796 cookies from firefox
[debug] Proxy map: {}
[debug] [RTBF] Extracting URL: https://www.rtbf.be/auvio/detail_la-belgique-criminelle?id=2920492
[RTBF] 2920492: Downloading webpage
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading MPD manifest
[RTBF] 2920492: Downloading m3u8 information
[RTBF] 2920492: Downloading ISM manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 2920492: Downloading 1 format(s): 5642+audio_fra-128
[debug] Invoking hlsnative downloader on "https://C9851Ec-az-westeurope.fsly.cdn.ebsd.ericsson.net/rtbf/auvio/assets/2920492_6BA97Bb/materials/hnUIW4LRM6_6BA97Bb/vod-idx.ism/vod-idx-video=5200000.m3u8"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 274
[download] Destination: La Belgique criminelle [2920492].f5642.mp4
[download]   2.3% of ~1016.19MiB at  824.23KiB/s ETA 10:38
```


Fixes #4443

pukkandan commented:
You can move all the redbee extractors into one file

Virusf commented:
I tested the modification and I had that
```shell
[debug] Command-line config: ['--verbose', '--cookies-from-browser', 'firefox', '--output', 'E:\\Youtube-DL\\Download\\%(title)s.%(ext)s', 'https://www.rtbf.be/auvio/detail_formule-1-grand-prix-de-france?id=2920922', '--check-formats']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05ef6]
[debug] Lazy loading extractors is disabled
[debug] Python 3.7.7 (CPython 64bit) - Windows-10-10.0.22000-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.4-full_build-www.gyan.dev (setts), ffprobe 5.1-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, secretstorage-3.3.1, sqlite3-2.6.0, websockets-10.1
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "C:\Users\virus\AppData\Roaming\Mozilla\Firefox\Profiles\97vujq2h.default-release\cookies.sqlite"
[Cookies] Extracted 2791 cookies from firefox
[debug] Proxy map: {}
[debug] [RTBF] Extracting URL: https://www.rtbf.be/auvio/detail_formule-1-grand-prix-de-france?id=2920922
[RTBF] 2920922: Downloading webpage
[RTBF] 2920922: Downloading webpage
[RTBF] 2920922: Downloading JSON metadata
[RTBF] 2920922: Downloading JSON metadata
[RTBF] 2920922: Downloading JSON metadata
[RTBF] 2920922: Downloading MPD manifest
[RTBF] 2920922: Downloading m3u8 information
[RTBF] 2920922: Downloading ISM manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] Testing format 1826
[hlsnative] Downloading m3u8 manifest
ERROR: This video is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format
  File "E:\Python3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "E:\Python3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "E:\Python3\lib\site-packages\yt_dlp\__main__.py", line 17, in <module>
    yt_dlp.main()
  File "E:\Python3\lib\site-packages\yt_dlp\__init__.py", line 940, in main
    _exit(*variadic(_real_main(argv)))
  File "E:\Python3\lib\site-packages\yt_dlp\__init__.py", line 932, in _real_main
    return ydl.download(all_urls)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 3298, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 3273, in wrapper
    res = func(*args, **kwargs)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 1424, in extract_info
    return self.__extract_info(url, self.get_info_extractor(ie_key), download, extra_info, process)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 1433, in wrapper
    return func(self, *args, **kwargs)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 1530, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 1587, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2631, in process_video_result
    or all(f.get('vcodec') == 'none' and f.get('acodec') != 'none' for f in formats)),
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2235, in final_selector
    return selector_function(ctx_copy)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2147, in selector_function
    picked_formats = list(f(ctx))
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2156, in selector_function
    for pair in itertools.product(selector_1(ctx), selector_2(ctx)):
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2225, in selector_function
    yield matches[format_idx - 1]
  File "E:\Python3\lib\site-packages\yt_dlp\utils.py", line 2705, in __getitem__
    self._cache.extend(itertools.islice(self._iterable, n))
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2128, in _check_formats
    yield from self._check_formats(formats)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 1899, in _check_formats
    success, _ = self.dl(temp_file.name, f, test=True)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 2876, in dl
    return fd.download(name, new_info, subtitle)
  File "E:\Python3\lib\site-packages\yt_dlp\downloader\common.py", line 444, in download
    ret = self.real_download(filename, info_dict)
  File "E:\Python3\lib\site-packages\yt_dlp\downloader\hls.py", line 83, in real_download
    'This video is DRM protected; Try selecting another format with --format or '
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 973, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "E:\Python3\lib\site-packages\yt_dlp\YoutubeDL.py", line 905, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())
    
```

pukkandan commented:
> ERROR: This video is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format

maybe actually read the error message? If all formats give this, yt-dlp cannot download that video

Virusf commented:
Yes all format gives the same error message
But after on the old video there was Widevine on it and it happened to the Download.
This is why I posted the error message in case

elyse0 commented:
I've added support for username and password, and added some tests n-n

I changed the extraction title because it was extracting a too generic name, for example:
**Title:** La Belgique Criminelle
**Subtitle:** 04 - Le crime de la rue Royale

Virusf commented:
I think it's dead

```shell
[debug] Command-line config: ['--verbose', '--username', 'PRIVATE', '--password', 'PRIVATE', '--output', 'E:\\Youtube-DL\\Download\\%(title)s.%(ext)s', 'https://www.rtbf.be/auvio/detail_formule-1-le-grand-prix?id=2923852', '--check-formats']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05ef6]
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Python 3.7.7 (CPython 64bit) - Windows-10-10.0.22000-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.1-full_build-www.gyan.dev (setts), ffprobe 5.1-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, secretstorage-3.3.1, sqlite3-2.6.0, websockets-10.1
[debug] Proxy map: {}
[RTBF] Downloading JSON metadata
[debug] [RTBF] Extracting URL: https://www.rtbf.be/auvio/detail_formule-1-le-grand-prix?id=2923852
[RTBF] 2923852: Downloading webpage
[RTBF] 2923852: Downloading JSON metadata
[RTBF] 2923852: Downloading JSON metadata
[RTBF] 2923852: Downloading JSON metadata
[RTBF] 2923852: Downloading MPD manifest
[RTBF] 2923852: Downloading m3u8 information
[RTBF] 2923852: Downloading ISM manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] Testing format 1826
[hlsnative] Downloading m3u8 manifest
ERROR: This video is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__main__.py", line 17, in <module>
    yt_dlp.main()
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 940, in main
    _exit(*variadic(_real_main(argv)))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 932, in _real_main
    return ydl.download(all_urls)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3298, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3273, in wrapper
    res = func(*args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1424, in extract_info
    return self.__extract_info(url, self.get_info_extractor(ie_key), download, extra_info, process)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1433, in wrapper
    return func(self, *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1530, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1587, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2631, in process_video_result
    or all(f.get('vcodec') == 'none' and f.get('acodec') != 'none' for f in formats)),
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2235, in final_selector
    return selector_function(ctx_copy)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2147, in selector_function
    picked_formats = list(f(ctx))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2156, in selector_function
    for pair in itertools.product(selector_1(ctx), selector_2(ctx)):
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2225, in selector_function
    yield matches[format_idx - 1]
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\utils.py", line 2705, in __getitem__
    self._cache.extend(itertools.islice(self._iterable, n))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2128, in _check_formats
    yield from self._check_formats(formats)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1899, in _check_formats
    success, _ = self.dl(temp_file.name, f, test=True)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2876, in dl
    return fd.download(name, new_info, subtitle)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\common.py", line 444, in download
    ret = self.real_download(filename, info_dict)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\hls.py", line 83, in real_download
    'This video is DRM protected; Try selecting another format with --format or '
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 973, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 905, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())


[info] Unable to download format 1826. Skipping...
[info] Testing format 766
[hlsnative] Downloading m3u8 manifest
ERROR: This video is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__main__.py", line 17, in <module>
    yt_dlp.main()
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 940, in main
    _exit(*variadic(_real_main(argv)))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 932, in _real_main
    return ydl.download(all_urls)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3298, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3273, in wrapper
    res = func(*args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1424, in extract_info
    return self.__extract_info(url, self.get_info_extractor(ie_key), download, extra_info, process)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1433, in wrapper
    return func(self, *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1530, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1587, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2631, in process_video_result
    or all(f.get('vcodec') == 'none' and f.get('acodec') != 'none' for f in formats)),
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2235, in final_selector
    return selector_function(ctx_copy)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2147, in selector_function
    picked_formats = list(f(ctx))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2156, in selector_function
    for pair in itertools.product(selector_1(ctx), selector_2(ctx)):
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2225, in selector_function
    yield matches[format_idx - 1]
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\utils.py", line 2705, in __getitem__
    self._cache.extend(itertools.islice(self._iterable, n))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2128, in _check_formats
    yield from self._check_formats(formats)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1899, in _check_formats
    success, _ = self.dl(temp_file.name, f, test=True)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2876, in dl
    return fd.download(name, new_info, subtitle)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\common.py", line 444, in download
    ret = self.real_download(filename, info_dict)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\hls.py", line 83, in real_download
    'This video is DRM protected; Try selecting another format with --format or '
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 973, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 905, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())


[info] Unable to download format 766. Skipping...
[info] Testing format 448
[hlsnative] Downloading m3u8 manifest
ERROR: This video is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__main__.py", line 17, in <module>
    yt_dlp.main()
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 940, in main
    _exit(*variadic(_real_main(argv)))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 932, in _real_main
    return ydl.download(all_urls)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3298, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3273, in wrapper
    res = func(*args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1424, in extract_info
    return self.__extract_info(url, self.get_info_extractor(ie_key), download, extra_info, process)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1433, in wrapper
    return func(self, *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1530, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1587, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2631, in process_video_result
    or all(f.get('vcodec') == 'none' and f.get('acodec') != 'none' for f in formats)),
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2235, in final_selector
    return selector_function(ctx_copy)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2147, in selector_function
    picked_formats = list(f(ctx))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2156, in selector_function
    for pair in itertools.product(selector_1(ctx), selector_2(ctx)):
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2225, in selector_function
    yield matches[format_idx - 1]
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\utils.py", line 2705, in __getitem__
    self._cache.extend(itertools.islice(self._iterable, n))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2128, in _check_formats
    yield from self._check_formats(formats)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1899, in _check_formats
    success, _ = self.dl(temp_file.name, f, test=True)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2876, in dl
    return fd.download(name, new_info, subtitle)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\common.py", line 444, in download
    ret = self.real_download(filename, info_dict)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\hls.py", line 83, in real_download
    'This video is DRM protected; Try selecting another format with --format or '
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 973, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 905, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())


[info] Unable to download format 448. Skipping...
[info] Testing format 130
[hlsnative] Downloading m3u8 manifest
ERROR: This video is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__main__.py", line 17, in <module>
    yt_dlp.main()
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 940, in main
    _exit(*variadic(_real_main(argv)))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 932, in _real_main
    return ydl.download(all_urls)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3298, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3273, in wrapper
    res = func(*args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1424, in extract_info
    return self.__extract_info(url, self.get_info_extractor(ie_key), download, extra_info, process)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1433, in wrapper
    return func(self, *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1530, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1587, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2631, in process_video_result
    or all(f.get('vcodec') == 'none' and f.get('acodec') != 'none' for f in formats)),
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2235, in final_selector
    return selector_function(ctx_copy)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2147, in selector_function
    picked_formats = list(f(ctx))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2156, in selector_function
    for pair in itertools.product(selector_1(ctx), selector_2(ctx)):
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2225, in selector_function
    yield matches[format_idx - 1]
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\utils.py", line 2705, in __getitem__
    self._cache.extend(itertools.islice(self._iterable, n))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2128, in _check_formats
    yield from self._check_formats(formats)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1899, in _check_formats
    success, _ = self.dl(temp_file.name, f, test=True)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2876, in dl
    return fd.download(name, new_info, subtitle)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\common.py", line 444, in download
    ret = self.real_download(filename, info_dict)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\hls.py", line 83, in real_download
    'This video is DRM protected; Try selecting another format with --format or '
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 973, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 905, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())


[info] Unable to download format 130. Skipping...
[info] Testing format audio-aacl-128-Français
[hlsnative] Downloading m3u8 manifest
ERROR: This video is DRM protected; Try selecting another format with --format or add --check-formats to automatically fallback to the next best format
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__main__.py", line 17, in <module>
    yt_dlp.main()
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 940, in main
    _exit(*variadic(_real_main(argv)))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\__init__.py", line 932, in _real_main
    return ydl.download(all_urls)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3298, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 3273, in wrapper
    res = func(*args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1424, in extract_info
    return self.__extract_info(url, self.get_info_extractor(ie_key), download, extra_info, process)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1433, in wrapper
    return func(self, *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1530, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1587, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2631, in process_video_result
    or all(f.get('vcodec') == 'none' and f.get('acodec') != 'none' for f in formats)),
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2235, in final_selector
    return selector_function(ctx_copy)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2147, in selector_function
    picked_formats = list(f(ctx))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2156, in selector_function
    for pair in itertools.product(selector_1(ctx), selector_2(ctx)):
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2136, in selector_function
    yield from f(ctx)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2225, in selector_function
    yield matches[format_idx - 1]
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\utils.py", line 2705, in __getitem__
    self._cache.extend(itertools.islice(self._iterable, n))
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2128, in _check_formats
    yield from self._check_formats(formats)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1899, in _check_formats
    success, _ = self.dl(temp_file.name, f, test=True)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2876, in dl
    return fd.download(name, new_info, subtitle)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\common.py", line 444, in download
    ret = self.real_download(filename, info_dict)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\downloader\hls.py", line 83, in real_download
    'This video is DRM protected; Try selecting another format with --format or '
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 973, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 905, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())


[info] Unable to download format audio-aacl-128-Français. Skipping...
ERROR: [RTBF] 2923852: Requested format is not available. Use --list-formats for a list of available formats
Traceback (most recent call last):
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1433, in wrapper
    return func(self, *args, **kwargs)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1530, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 1587, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "E:\Youtube-DL\yt-dlp_rtbf\yt_dlp\YoutubeDL.py", line 2642, in process_video_result
    expected=True, video_id=info_dict['id'], ie=info_dict['extractor'])
yt_dlp.utils.ExtractorError: [RTBF] 2923852: Requested format is not available. Use --list-formats for a list of available formats
```

pukkandan commented:
test my changes and let me know

pukkandan commented:
I converted RedBeeIE into a base class since I didn't see any reason why it should have a valid_url

elyse0 commented:
I've tested it and everything works n-n

I just added a line to sort formats by `extension`, because using Ism format is quiteee slow o-O

```bash
[Cookies] Extracting cookies from firefox
[Cookies] Extracted 2617 cookies from firefox
[RTBF] 2920492: Downloading webpage
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading MPD manifest
[RTBF] 2920492: Downloading m3u8 information
[RTBF] 2920492: Downloading ISM manifest
[info] Available formats for 2920492:
ID                      EXT  RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC        VBR ACODEC      ABR ASR MORE INFO
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
audio-aacl-128-Français m3u8 audio only     │                  m3u8  │ audio only        unknown             [fr] Français
audio_fra=128000        m4a  audio only     │ ~ 24.60MiB  128k dash  │ audio only        mp4a.40.2  128k 48k [fr] DASH audio, m4a_dash
130                     m4a  audio only     │ ~ 24.99MiB  130k m3u8  │ audio only        mp4a.40.2  128k
audio_fra-128           isma audio only     │ ~ 24.60MiB  128k ism   │ audio only        AACL       128k 48k
video-300               ismv 480x270        │ ~ 57.66MiB  300k ism   │ AVC1         300k video only
video=300000            mp4  480x270     25 │ ~ 57.66MiB  300k dash  │ avc1.4D4015  300k video only          DASH video, mp4_dash
448                     mp4  480x270     25 │ ~ 86.11MiB  448k m3u8  │ avc1.4D4015  448k video only
video-600               ismv 640x360        │ ~115.33MiB  600k ism   │ AVC1         600k video only
video=600000            mp4  640x360     25 │ ~115.33MiB  600k dash  │ avc1.4D401E  600k video only          DASH video, mp4_dash
766                     mp4  640x360     25 │ ~147.23MiB  766k m3u8  │ avc1.4D401E  766k video only
video-1600              ismv 960x540        │ ~307.54MiB 1600k ism   │ AVC1        1600k video only
video=1600000           mp4  960x540     25 │ ~307.54MiB 1600k dash  │ avc1.4D401F 1600k video only          DASH video, mp4_dash
1826                    mp4  960x540     25 │ ~350.98MiB 1826k m3u8  │ avc1.4D401F 1826k video only
video-3000              ismv 1280x720       │ ~576.64MiB 3000k ism   │ AVC1        3000k video only
video=3000000           mp4  1280x720    50 │ ~576.64MiB 3000k dash  │ avc1.4D4020 3000k video only          DASH video, mp4_dash
3310                    mp4  1280x720    50 │ ~636.22MiB 3310k m3u8  │ avc1.4D4020 3310k video only
video-5200              ismv 1920x1080      │ ~999.50MiB 5200k ism   │ AVC1        5200k video only
video=5200000           mp4  1920x1080   50 │ ~999.50MiB 5200k dash  │ avc1.4D402A 5200k video only          DASH video, mp4_dash
5642                    mp4  1920x1080   50 │ ~  1.06GiB 5642k m3u8  │ avc1.4D402A 5642k video only
```

### Using ism (+2min to download audio)
```bash
[Cookies] Extracting cookies from firefox
[Cookies] Extracted 2618 cookies from firefox
[RTBF] 2920492: Downloading webpage
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading MPD manifest
[RTBF] 2920492: Downloading m3u8 information
[RTBF] 2920492: Downloading ISM manifest
[info] 2920492: Downloading 1 format(s): 5642+audio_fra-128
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 274
[download] Destination: 04 - Le crime de la rue Royale [2920492].f5642.mp4
WARNING: The download speed shown is only of one thread. This is a known issue and patches are welcome
[download] 100% of 1003.93MiB in 02:59
[ism] Total fragments: 820
[download] Destination: 04 - Le crime de la rue Royale [2920492].faudio_fra-128.isma
[download] 100% of 24.16MiB in 02:55
[Merger] Merging formats into "04 - Le crime de la rue Royale [2920492].mp4"
```

### Using m3u8 (11s to download audio)
```bash
[Cookies] Extracting cookies from firefox
[Cookies] Extracted 2619 cookies from firefox
[RTBF] 2920492: Downloading webpage
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading JSON metadata
[RTBF] 2920492: Downloading MPD manifest
[RTBF] 2920492: Downloading m3u8 information
[RTBF] 2920492: Downloading ISM manifest
[info] 2920492: Downloading 1 format(s): 5642+130
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 274
[download] Destination: 04 - Le crime de la rue Royale [2920492].f5642.mp4
WARNING: The download speed shown is only of one thread. This is a known issue and patches are welcome
[download] 100% of 1003.93MiB in 03:35
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 274
[download] Destination: 04 - Le crime de la rue Royale [2920492].f130.m4a
WARNING: The download speed shown is only of one thread. This is a known issue and patches are welcome
[download] 100% of 24.05MiB in 00:11
[Merger] Merging formats into "04 - Le crime de la rue Royale [2920492].mp4"
````

pukkandan commented:
Sorting by ext is incorrect. Give the ism formats a `source_preference=-10` and then sort by `['res', 'source']`. Alternately, sorting by `['res', 'proto']` may work too

pukkandan commented:
Note: The order you see in `-F` is worst to best according to the given sorting

elyse0 commented:
Okay, it's fixed. As you said, sorting by 'res' and 'proto' also works n-n

jfou87 commented:
Hello elyse0, can you tell me how can I download and use the branch RTBF you done ? 
Sorry for my stupid question
